File: Concurrency/BadCounter.java
import java.util.ArrayList;
import java.util.stream.IntStream;

class Counter {
    int count;
    Counter() {
        this.count = 0;
    }
}

class TestTask implements Runnable {
    int s;
    Counter c;

    TestTask(int st, Counter c) {
        this.s = st;
        this.c = c;
    }

    @Override
    public void run() {
        for(int i = 0; i < this.s; i++){
            c.count += 1;
        }
    }
}

class ThreadManager {
    ArrayList<Thread> allThreads;

    ThreadManager() {
        this.allThreads = new ArrayList<Thread>();
    }

    void addThread(Runnable r) {
        Thread th = new Thread(r);
        allThreads.add(th);
    }

    void startThreads() {
        for(Thread th: this.allThreads) {
            th.start();
        }
    }

    void joinAll() throws InterruptedException {
        for(Thread th: this.allThreads) {
            th.join();
        }
    }

    void addMultipleThreads(int threadCount, int count, Counter c) {
        IntStream.range(0, threadCount)
                    .forEach(x -> {
                        TestTask t = new TestTask(count, c);
                        this.addThread(t);
                    });
    }
}

public class BadCounter {
    public static void main(String[] args) throws InterruptedException {
        System.out.println("Start Things");
        Counter c = new Counter();
        ThreadManager th = new ThreadManager();
        th.addMultipleThreads(1000, 10000, c);
        th.startThreads();
        th.joinAll();
        System.out.println("Done" + " " + c.count);
    }
}

File: Concurrency/badCounter.py
import threading
import time

class BadCounter():
    def __init__(self):
        self.threads = []
        self.counter = 0

    def counter1(self):
        for _ in range(1000):
            self.counter += 1
            

    def addThreads(self, th):
        self.threads.append(th)

    def startThreads(self):
        for th in self.threads:
            th.start()

    def wait(self):
        for th in self.threads:
            th.join()

    def createThreads(self, threadNumber):
        for _ in range(threadNumber):
            self.addThreads(threading.Thread(target=self.counter1))

    def testCounter(self):
        self.createThreads(100)

        self.startThreads()
        self.wait()

        print(self.counter)

bdc = BadCounter()
bdc.testCounter()